src/number/format.js

```diff
@@ -68,6 +68,32 @@ return function( number, properties ) {
 		number *= 1000;
 	}
 
+	var compactPattern, compactDigits, compactProperties, divisor, pluralForm, zeroes,
+	originalNumber;
+
+	// Compact mode: initial number digit processing
+	if ( compactMap ) {
+		originalNumber = number;
+		zeroes = Array( Math.floor( number ).toString().length ).join( "0" );
+		if ( zeroes.length >= 3 ) {
+
+			// use default plural form to perform initial decimal shift
+			compactPattern = compactMap[ "1" + zeroes + "-count-other" ];
```

Comment:
At this point (the format execution itself), `compactMap` data structure should be the most convenient possible for use here (prepared by format-properties stage), e.g.:

```
compactMap: {
  3: {
    one: "0 thousand",
    other: "0 thousand"
  },
  4: {one, other},
  ...
  15: {one, other}
}
```

That would eliminate the need of creating the zeros array and would simplify the key above.